### PR TITLE
Add a copy-state zebrad command, which copies blocks between two state services

### PR DESCRIPTION
**Pull request by teor2345**
## Motivation

As part of the `lightwalletd` work, we need to make significant changes to Zebra's on-disk database format. But syncing the entire chain takes 2-6 hours, which will make development very slow.

Instead, we can copy blocks from one state directory to another, skipping the network, structural validation, semantic validation, and most contextual validation. (But still confirming that the block data is bound to the block header, because that's required for NU5 anyway.)

Copying blocks between identical state formats takes about 15 minutes with a single-threaded copy.

This draft PR is part of planning the wallet support. We need to know how hard it will be to change the database format, and what we need to do to test those changes.

### Zebra APIs

Since `zebra-state` is a separate crate, and there we can start two state instances, and copy blocks between them.

## Solution

- Initialize `zebrad` as a long-running server command, with all the default components (tracing, metrics, etc.)
- Open two state services
- Copy blocks between them, logging progress at regular intervals
- Verify that the blocks were copied correctly

## Review

Anyone can review this low-priority PR.

### Reviewer Checklist

  - [x] Code makes sense

This is a draft dev-only command, so it doesn't need acceptance tests yet. (We might want to add some when we've used it a bit, and its options are stable.)

## Follow Up Work

- Actually modify the database format
- Modify this command to copy between two different database formats

### CODE REVIEW

**Comment by teor2345** [comment]
This PR needs to be fixed as part of ticket #3133, and it also provides a convenient way to test PR #3133.

So I've upgraded its priority to medium.

File: zebrad/src/commands.rs [hunk 1]
```diff
@@ -21,6 +25,10 @@ pub const CONFIG_FILE: &str = "zebrad.toml";
 /// Zebrad Subcommands
 #[derive(Command, Debug, Options)]
 pub enum ZebradCmd {
+    /// The `copy-state` subcommand, used to debug cached chain state
+    #[options(help = "copy cached chain state (debug only)")]
+    CopyState(CopyStateCmd),
```
Comment (dconnolly) [inline zebrad/src/commands.rs]: I wonder if we can feature-gate commands like this, so it does not get compiled into `zebrad` by default

File: zebrad/src/commands/copy_state.rs [hunk 2]
```diff
@@ -0,0 +1,437 @@
...
+//!    * fetches blocks from the best finalized chain from permanent storage,
+//!      in the old format
+//!  * Target New State Service
+//!    * writes best finalized chain blocks to permanent storage,
+//!      in the new format
+//!    * only performs essential contextual verification of blocks,
+//!      to make sure that block data hasn't been corrupted by
+//!      receiving blocks in the new format
+//!    * fetches blocks from the best finalized chain from permanent storage,
+//!      in the new format
```
Comment (dconnolly) [inline zebrad/src/commands/copy_state.rs:34]: 😍

File: zebrad/src/commands/copy_state.rs [hunk 3]
```diff
@@ -0,0 +1,437 @@
...
+use std::{cmp::min, path::PathBuf};
+
+use abscissa_core::{config, Command, FrameworkError, Options, Runnable};
+use color_eyre::eyre::{eyre, Report};
+
+use tokio::time::Instant;
+use tower::{Service, ServiceExt};
+use zebra_chain::{block::Height, parameters::Network};
+use zebra_state as old_zs;
+use zebra_state as new_zs;
```
Comment (dconnolly) [inline zebrad/src/commands/copy_state.rs:45]: 🤯

File: zebrad/src/commands/copy_state.rs [hunk 4]
```diff
@@ -0,0 +1,437 @@
...
+
+use crate::{
+    components::tokio::{RuntimeRun, TokioComponent},
+    config::ZebradConfig,
+    prelude::*,
+    BoxError,
+};
+
+/// How often we log info-level progress locks
+const PROGRESS_HEIGHT_INTERVAL: u32 = 5_000;
```
Comment (dconnolly) [inline zebrad/src/commands/copy_state.rs:55]: 👍

File: zebrad/src/commands/copy_state.rs [hunk 5]
```diff
@@ -0,0 +1,437 @@
...
+
+/// `copy-state` subcommand
+#[derive(Command, Debug, Options)]
+pub struct CopyStateCmd {
+    /// Source height that the copy finishes at.
+    #[options(help = "stop copying at this source height")]
+    max_source_height: Option<u32>,
+
```
Comment (dconnolly) [inline zebrad/src/commands/copy_state.rs:63]: 👍

File: zebrad/src/commands/copy_state.rs [hunk 6]
```diff
@@ -0,0 +1,437 @@
...
+        let source_tip = match source_tip {
+            old_zs::Response::Tip(Some(source_tip)) => source_tip,
+            old_zs::Response::Tip(None) => Err("empty source state: no blocks to copy")?,
+
+            response => Err(format!(
+                "unexpected response to Tip request: {:?}",
+                response,
+            ))?,
+        };
+        let source_tip_height = source_tip.0 .0;
```
Comment (dconnolly) [inline zebrad/src/commands/copy_state.rs:148]: This is odd, is this a GitHub render artifact?

image

File: zebrad/src/commands/copy_state.rs [hunk 7]
```diff
@@ -0,0 +1,437 @@
...
+                     target data hash: {:?},\n \
+                     source block: {:?},\n \
+                     target block: {:?}",
+                    max_copy_height,
+                    source_block_hash,
+                    target_block_commit_hash,
+                    target_block_data_hash,
+                    source_block,
+                    target_block,
+                ))?;
```
Comment (dconnolly) [inline zebrad/src/commands/copy_state.rs:292]: 👍

File: zebrad/src/commands/copy_state.rs [hunk 8]
```diff
@@ -0,0 +1,437 @@
...
+        let final_target_tip = match final_target_tip {
+            new_zs::Response::Tip(Some(target_tip)) => target_tip,
+            new_zs::Response::Tip(None) => Err("empty target state: expected written blocks")?,
+
+            response => Err(format!(
+                "unexpected response to Tip request: {:?}",
+                response,
+            ))?,
+        };
+        let final_target_tip_height = final_target_tip.0 .0;
```
Comment (dconnolly) [inline zebrad/src/commands/copy_state.rs:326]: Funny:

image

**Review by dconnolly — DISMISSED** [review]
Looking great, just some small points

File: zebrad/src/commands.rs (hunk 1 above)
Comment (teor2345) [reply zebrad/src/commands.rs]: There are a few different ways to implement this, and they depend on work we haven't done yet. So I opened ticket #3279 to make this change.

File: zebrad/src/commands/copy_state.rs (hunk 6 above)
Comment (teor2345) [reply zebrad/src/commands/copy_state.rs:148]: It's actually a `rustfmt` feature, to work around a `rustc` bug. The `rustc` bug was fixed in 2020, and the `rustfmt` rules were changed in version 2.

I opened ticket #3280 to fix this.

File: zebrad/src/commands/copy_state.rs (hunk 8 above)
Comment (teor2345) [reply zebrad/src/commands/copy_state.rs:326]: Will be fixed by #3280

File: zebrad/src/commands/copy_state.rs (hunk 3 above)
Comment (teor2345) [reply zebrad/src/commands/copy_state.rs:45]: Yep, it just works, with whatever state versions we like.

If we're using two different `zebra` checkouts, we'll also need a `new_zc` and `old_zc`. And we'll need to turn blocks into bytes when copying and comparing them. But we can make those changes if we ever need to use two different state versions.

Comment (teor2345) [reply zebrad/src/commands/copy_state.rs:45]: If `zebra-chain` hasn't changed, we can just use two `zebra-state` directories based on the same `zebra-chain`. And the current `copy-state` code should work fine.

**Comment by teor2345** [comment]
I opened tickets for the issues you raised, some of them depend on other changes. So I'll just go ahead and merge this now.

**Review by dconnolly — APPROVED** [review]

**Pull request merged**